src/object/cli_obj.c

```diff
@@ -2748,3 +2792,93 @@ dc_obj_query_key(tse_task_t *api_task)
 
 	return rc;
 }
+
+static int
+shard_sync_prep(struct shard_auxi_args *shard_auxi, struct dc_object *obj,
+		struct obj_auxi_args *obj_auxi, uint64_t dkey_hash,
+		uint32_t grp_idx)
+{
+	struct daos_obj_sync	*obj_args;
+	struct shard_sync_args	*shard_args;
+
+	obj_args = dc_task_get_args(obj_auxi->obj_task);
+	shard_args = container_of(shard_auxi, struct shard_sync_args, sa_auxi);
+	shard_args->sa_epoch = obj_args->epoch[grp_idx];
+
+	return 0;
+}
+
+int
+dc_obj_sync(tse_task_t *task)
+{
+	struct daos_obj_sync	*args;
+	struct obj_auxi_args	*obj_auxi = NULL;
+	struct dc_object	*obj = NULL;
+	daos_epoch_t		 epoch;
+	uint32_t		 replicas;
+	uint32_t		 map_ver;
+	int			 rc;
+	int			 i;
+
+	if (srv_io_mode != DIM_DTX_FULL_ENABLED)
+		D_GOTO(out_task, rc = 0);
+
+	args = dc_task_get_args(task);
+	D_ASSERTF(args != NULL,
+		  "Task Argument OPC does not match DC OPC\n");
+
+	rc = obj_req_valid(args, DAOS_OBJ_RPC_SYNC, &epoch);
+	if (rc != 0)
+		D_GOTO(out_task, rc);
+
+	D_ASSERT(epoch);
+
+	obj = obj_hdl2ptr(args->oh);
+	if (obj == NULL)
+		D_GOTO(out_task, rc = -DER_NO_HDL);
+
+	rc = obj_ptr2pm_ver(obj, &map_ver);
+	if (rc != 0) {
+		obj_decref(obj);
+		D_GOTO(out_task, rc);
+	}
+
+	rc = obj_reg_comp_cb(task, DAOS_OBJ_RPC_SYNC, map_ver, &obj_auxi,
+			     &obj, sizeof(obj));
+	if (rc != 0) {
+		obj_decref(obj);
+		D_GOTO(out_task, rc);
+	}
+
+	/* Need to mark sync epoch on server, so even if the @replicas is 1,
+	 * we still need to send SYNC RPC to the server.
+	 */
+	replicas = obj_get_replicas(obj);
+	if (!obj_auxi->io_retry) {
+		*args->nr = obj->cob_shards_nr / replicas;
+		D_ALLOC_ARRAY(*args->epoch, *args->nr);
+		if (*args->epoch == NULL)
+			D_GOTO(out_task, rc = -DER_NOMEM);
```

Comment:
so the caller of dc_obj_sync_task_create() should free it when the task succeed?
but if the task failed should it be freed internally in obj_comp_cb()?

Reply:
You are right. I will fix it.